core/src/main/java/com/digitalpebble/stormcrawler/util/MetadataKryoSerializer.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Licensed to DigitalPebble Ltd under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership.
+ * DigitalPebble licenses this file to You under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * <p>http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.digitalpebble.stormcrawler.util;
+
+import static java.util.Collections.EMPTY_MAP;
+
+import com.digitalpebble.stormcrawler.Metadata;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.Serializer;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.serializers.DefaultArraySerializers;
+import com.esotericsoftware.kryo.serializers.DefaultSerializers;
+import java.util.Map;
+import org.apache.storm.serialization.types.HashMapSerializer;
+
+/** A Kryo serializer for Metadata */
+// We need that serializer to make sure, that we do protect an empty metadata instance.
+// Otherwise it'll be converted to an HashMap.
+public class MetadataKryoSerializer extends Serializer<Metadata> {
```

Comment:
Do we need to explicitly register this serializer to be used in the context of a `flux`-based topology? If so, we would need to adjust some config files.

Reply:
It is not explicitly registered (at least when I understand the term "explicit" in the kryo-context). That serializer is set as default serializer via annotation at the `Metadata`-class. 
So flux shouldn't have a problem with that. Even the registration in the topology is the same as before: see core/src/main/java/com/digitalpebble/stormcrawler/ConfigurableTopology.java

And we need that serializer, otherwise the EMPTY_MAP wont be handled correctly, and then we get that what @jnioche had to correct with the "remove empty metadata from ...."

Reply:
> It is not explicitly registered (at least what I understand by the term "explicit" in the kryo-context). That serializer is set as default serializer via annotation at the `Metadata`-class. So flux shouldn't have a problem with that. Even the registration in the topology is the same as before: see core/src/main/java/com/digitalpebble/stormcrawler/ConfigurableTopology.java
> 
> And we need that serializer, otherwise the EMPTY_MAP wont be handled correctly, and then we get that what @jnioche had to correct with the "remove empty metadata from ...."

Ups, didn't mean to write the @ that autocompletion got me off guard. :'D